Introduce a new native method to access the master key

fzakaria opened the pull request:
This change accompanies the broader pull-request introduced by 
https://github.com/netty/netty/pull/8653 to allow access to the SSL session master key for debugging purposes.

Accessing the master key is useful if you ever need to decrypt the TLS session for debugging purposes with tools such as Wireshark.

Access of the field from the SSL session struct is direct as the wrapper method was only introduced in 1.1.0+  https://www.openssl.org/docs/man1.1.0/ssl/SSL_SESSION_get_master_key.html

normanmaurer commented:
@netty-bot test this please

normanmaurer commented on openssl-dynamic/src/main/c/ssl.c:
```diff
@@ -1607,6 +1607,54 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
+#if OPENSSL_VERSION_NUMBER < 0x10100000L
+/*
+  Backport of SSL_SESSION_get_master_key from 1.1
+*/
```
nit: use either `//` or:

```
/*
 * 
 */
```

normanmaurer commented on openssl-dynamic/src/main/c/ssl.c:
```diff
@@ -1607,6 +1607,54 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
+size_t SSL_SESSION_get_master_key(const SSL_SESSION *session,
+                                  unsigned char *out, size_t outlen)
+{
+    if (outlen == 0)
```
please use `{...}`

normanmaurer commented on openssl-dynamic/src/main/c/ssl.c:
```diff
@@ -1607,6 +1607,54 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
+{
+    if (outlen == 0)
+        return session->master_key_length;
+    if (outlen > session->master_key_length)
```
please use `{...}`

normanmaurer commented on openssl-dynamic/src/main/c/ssl.c:
```diff
@@ -1607,6 +1607,54 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
+    return outlen;
+}
+#endif
+
```
nit: remove one empty line

normanmaurer commented on openssl-dynamic/src/main/c/ssl.c:
```diff
@@ -1607,6 +1607,54 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
+    SSL_SESSION *session = NULL;
+    if (session == NULL) {
+        return NULL;
+    }
```
This does not look correct.. session will also be `NULL` here. Did you forget to call `SSL_get0_SESSION(...)` ?

normanmaurer commented on openssl-dynamic/src/main/c/ssl.c:
```diff
@@ -1607,6 +1607,54 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
+    return jKey;
+}
+
+
```
nit: remove one empty line

fzakaria commented:
@normanmaurer Thanks for the comments. I'll address the feedback. I'd like to include a unit test, do you have a good example of how to start a TLS session with the code in this JAR? 

I'd only know how to write a test by starting client+server but that would need a cyclic dependency on SslHandler in Netty.

normanmaurer commented:
@fzakaria unfortually no :( I guess the best bet is to include a test in Netty itself.

fzakaria commented:
Thanks. I'll work on that for a post PR commit I guess. 
A good test would be if in Netty itself I could also keep a copy of the encrypted messages before the engine grabs it and then manually decrypt it as well.

fzakaria commented:
@normanmaurer Addressed your feedback -- 
I really wanted to include a test for posterity but I've been stuck on getting it to work.

I've packaged the work on the test case so far:
https://gist.github.com/fzakaria/f1de28943c8932280254dd90ffbc4ebd

fzakaria commented:
@normanmaurer I've included a test in **netty** that proves the methods work by manually decrypting the TLS application data -- https://github.com/netty/netty/pull/8699

fzakaria commented:
@normanmaurer *ping*

normanmaurer commented:
@fzakaria sorry for the delay... Will try to review tomorrow latest.

normanmaurer commented on openssl-dynamic/src/main/c/ssl.c:
```diff
@@ -1607,6 +1607,137 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
+    TCN_ASSERT(bytesMoved == keyLength);
+
+    jbyteArray jKey = (*e)->NewByteArray(e, (jsize) bytesMoved);
+    (*e)->SetByteArrayRegion(e, jKey, 0, bytesMoved, (jbyte*) key);
```
I think we need to guard against when `NewByteArray(...)` returns `NULL` (like if there is no memory left).

fzakaria replied:
I can add the NULL checks however I stuck to previous convention where the other JNI calls did no such guard.

normanmaurer commented on openssl-dynamic/src/main/c/ssl.c:
```diff
@@ -1607,6 +1607,137 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
+    size_t bytesMoved = SSL_get_server_random(ssl_, key, keyLength);
+    TCN_ASSERT(bytesMoved == keyLength);
+
+    jbyteArray jKey = (*e)->NewByteArray(e, (jsize) bytesMoved);
```
same as above.

normanmaurer commented on openssl-dynamic/src/main/c/ssl.c:
```diff
@@ -1607,6 +1607,137 @@ TCN_IMPLEMENT_CALL(jboolean, SSL, setCipherSuites)(TCN_STDARGS, jlong ssl,
+    size_t bytesMoved = SSL_SESSION_get_master_key(session, key, keyLength);
+    TCN_ASSERT(bytesMoved == keyLength);
+
+    jbyteArray jKey = (*e)->NewByteArray(e, (jsize) bytesMoved);
```
same as above

fzakaria commented:
@normanmaurer @Lukasa  Please see https://github.com/netty/netty/pull/8699/files#diff-7a4cbf2f526e2066184fb9ef098cd815R1127 also on a test-case of it being used to then decrypt the TLS data by hand.

fzakaria commented:
Done.

normanmaurer commented:
@fzakaria can you do me a favour and squash + adjust the commit message to match our template:

https://netty.io/wiki/writing-a-commit-message.html

normanmaurer commented:
@fzakaria thanks

fzakaria commented:
Thanks for merging @normanmaurer 
As soon as new version is released -- I will bump the _pom.xml_ in the netty pull-request and it should be _good-to-go_ as well.
